<h2>Download and Clean Dataset</h2>



Let's start by importing the pandas and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

<h4>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. 
Ingredients include:</h4>

<h4>1. Cement</h4>

<h4>2. Blast Furnace Slag</h4>

<h4>3. Fly Ash</h4>

<h4>4. Water</h4>

<h4>5. Superplasticizer</h4>

<h4>6. Coarse Aggregate</h4>

<h4>7. Fine Aggregate</h4>

Let's read the dataset into a pandas dataframe.

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 
162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine 
aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

<h4>Let's check how many data points we have.</h4>

In [3]:
    concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

<h4>Split data into predictors and target</h4>


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<h2>Import Keras</h2>


Let's go ahead and import the Keras library

In [11]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [12]:

from keras.models import Sequential
from keras.layers import Dense

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. 
It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [14]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [15]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

<h2>Train and Test the Network</h2>


Let's call the function now to create our model.

In [16]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.

In [17]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 0s - loss: 1573.5137
Epoch 2/50
23/23 - 0s - loss: 1557.6000
Epoch 3/50
23/23 - 0s - loss: 1542.1434
Epoch 4/50
23/23 - 0s - loss: 1526.2441
Epoch 5/50
23/23 - 0s - loss: 1510.3969
Epoch 6/50
23/23 - 0s - loss: 1493.4929
Epoch 7/50
23/23 - 0s - loss: 1476.0061
Epoch 8/50
23/23 - 0s - loss: 1457.3730
Epoch 9/50
23/23 - 0s - loss: 1437.7982
Epoch 10/50
23/23 - 0s - loss: 1416.7296
Epoch 11/50
23/23 - 0s - loss: 1394.6315
Epoch 12/50
23/23 - 0s - loss: 1370.9712
Epoch 13/50
23/23 - 0s - loss: 1345.7916
Epoch 14/50
23/23 - 0s - loss: 1318.8616
Epoch 15/50
23/23 - 0s - loss: 1290.8319
Epoch 16/50
23/23 - 0s - loss: 1260.9135
Epoch 17/50
23/23 - 0s - loss: 1229.6317
Epoch 18/50
23/23 - 0s - loss: 1197.0483
Epoch 19/50
23/23 - 0s - loss: 1163.6628
Epoch 20/50
23/23 - 0s - loss: 1128.2058
Epoch 21/50
23/23 - 0s - loss: 1093.0044
Epoch 22/50
23/23 - 0s - loss: 1055.9819
Epoch 23/50
23/23 - 0s - loss: 1018.9786
Epoch 24/50
23/23 - 0s - loss: 981.5669
Epoch 25/50
23/23 - 0s - l

Next we need to evaluate the model on the test data.

In [18]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 599us/step - loss: 264.0535


264.0534973144531

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [19]:
from sklearn.metrics import mean_squared_error

In [20]:
    mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

264.053513271446 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [21]:
total_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 112.2741928100586
MSE 2: 94.7568130493164
MSE 3: 73.75532531738281
MSE 4: 75.20570373535156
MSE 5: 75.76649475097656
MSE 6: 78.99966430664062
MSE 7: 81.0528335571289
MSE 8: 67.5251693725586
MSE 9: 71.1544189453125
MSE 10: 64.4734878540039
MSE 11: 66.81637573242188
MSE 12: 52.893672943115234
MSE 13: 59.396270751953125
MSE 14: 51.70122146606445
MSE 15: 46.61565017700195
MSE 16: 37.90452575683594
MSE 17: 44.05173110961914
MSE 18: 41.93095397949219
MSE 19: 40.54659652709961
MSE 20: 42.82798767089844
MSE 21: 40.36758041381836
MSE 22: 40.53774642944336
MSE 23: 37.714630126953125
MSE 24: 39.75986862182617
MSE 25: 42.97623825073242
MSE 26: 42.02021408081055
MSE 27: 39.499656677246094
MSE 28: 40.4639778137207
MSE 29: 44.572715759277344
MSE 30: 41.97931671142578
MSE 31: 39.616424560546875
MSE 32: 36.623966217041016
MSE 33: 36.64358901977539
MSE 34: 43.810123443603516
MSE 35: 40.92860794067383
MSE 36: 45.64309310913086
MSE 37: 39.58758544921875
MSE 38: 43.60523986816406
MSE 39: 40.28958892